### ISSUE TITLE
InSet filter fails in query containing categorical dataset

### ISSUE THREAD

**Issue by lachlanhurst** [vpac-innovations/rsa#30]
Queries run with a categorical dataset produce a validation exception.

The following query was generated by the LCM interface.

``` xml
<query>
  <input id="LandUse_9" href="rsa:LandUse/100m"/>
  <input id="boundaries_10" href="rsa:boundaries/100m"/>
  <output id="output">
    <grid ref="#LandUse_9"/>
    <variable name="mask" ref="#SA3/output"/>
  </output>
  <filter id="set_11" cls="org.vpac.ndg.query.InSet">
    <literal name="ids" value="1,373"/>
    <sampler name="input" ref="#LandUse_9/LandUse"/>
  </filter>
  <filter id="LGA" cls="org.vpac.ndg.query.stats.Categories">
    <literal name="buckets" value="categorical"/>
    <sampler name="input" ref="#set_11/output"/>
    <sampler name="categories" ref="#boundaries_10/LGA"/>
  </filter>
  <filter id="SA3" cls="org.vpac.ndg.query.stats.Categories">
    <literal name="buckets" value="categorical"/>
    <sampler name="input" ref="#LGA/output"/>
    <sampler name="categories" ref="#boundaries_10/SA3"/>
  </filter>
</query>
```

When run, the query produced the following stack trace.

```
[ERROR] [07/22/2014 16:00:52.974] [Workers-akka.actor.default-dispatcher-9] [akka://Workers/user/worker/exec] Field #set_11/input has not been attached to an input.
org.vpac.ndg.query.QueryConfigurationException: Field #set_11/input has not been attached to an input.
    at org.vpac.ndg.query.GroupImpl.getValue(GroupImpl.java:88)
    at org.vpac.ndg.query.GroupImpl.add(GroupImpl.java:72)
    at org.vpac.ndg.query.FilterAdapter.gatherInputConstraints(FilterAdapter.java:292)
    at org.vpac.ndg.query.FilterFactory.createFilter(FilterFactory.java:119)
    at org.vpac.ndg.query.FilterFactory.createFilters(FilterFactory.java:80)
    at org.vpac.ndg.query.Query.setMemento(Query.java:190)
    at org.vpac.worker.WorkExecutor.executeQuery(WorkExecutor.java:183)
    at org.vpac.worker.WorkExecutor.onReceive(WorkExecutor.java:100)
    at akka.actor.UntypedActor$$anonfun$receive$1.applyOrElse(UntypedActor.scala:167)
    at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
    at akka.actor.UntypedActor.aroundReceive(UntypedActor.scala:97)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
    at akka.actor.ActorCell.invoke(ActorCell.scala:487)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:238)
    at akka.dispatch.Mailbox.run(Mailbox.scala:220)
    at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by z0u** [comment]
Working on a fix in fix-filter-inset branch.

**Comment by z0u** [comment]
The problem is that the `InSet` filter extends `InRange`, but it mistakenly hides (not overrides) some of its fields. It's probably unwise to have filters inherit from each other at all. But in the fix so far, I've split the common functionality out into an abstract base class.

It should work but now I'm getting other weird errors: when I run a query, it fails because the output file already exists. I thought the worker would create a unique temp directory, but maybe not.

Anyway, this may now be fixed but I need to test further.

**Comment by lachlanhurst** [comment]
Stepping through the code I get the actual source of the exception as `org.vpac.ndg.query.QueryConfigurationException: Public filter field set_5.lower is unbound.` , but yes, this does manifest as some weird output file exists error.

Also, the worker just keeps trying, and just keeps failing to process the job.

**Comment by lachlanhurst** [comment]
My guess is that the query validation is still trying to validate having an upper and lower bound as the `InSet` class inherits from `InRange`.

Tempted to change [this line](https://github.com/vpac-innovations/rsa/blob/akka-workers/src/rsaquery/src/org/vpac/ndg/query/FilterAdapter.java#L474), but it's close enough to home time for me... 

Maybe a simple matter of [SO: getFields vs getDeclaredFields](http://stackoverflow.com/questions/16966629/what-is-the-difference-between-getfields-getdeclaredfields-in-java-reflection) ?

**Comment by z0u** [comment]
We need to use `getFields` because otherwise it would be impossible for a filter to extend another one: inputs could not be assigned to the sub-classes. We could just disallow subclassing, but I'd prefer to just be clear in the documentation about how to subclass sensibly.

It turns out there was a small error in my patch: `InSet` was still inheriting from the non-abstract `InRange` filter, which declares its own fields that aren't set by the query. I've changed it to inherit from `In` and now it works.

Fixed in b22609d581a5db467a5003faa46c315007007b1d, but I haven't merged it back into `akka-workers` yet; I want to add a bit of error checking first to catch these mistakes. But you should be able to use that commit if you want to.

![screenshot from 2014-07-23 08 51 17](https://cloud.githubusercontent.com/assets/1408875/3666529/bff445bc-11f2-11e4-83d8-3ae6df029482.png)
